lib/rules/padding-between-test-functions.js

```diff
@@ -0,0 +1,166 @@
+        var lastToken = sourceCode.getLastToken(node),
+        tokenAfterLast = sourceCode.getTokenAfter(lastToken, { includeComments: true }),
+        blockHasBottomPadding = isPaddingBetweenTokens(lastToken, tokenAfterLast),
+        errors = [];
+
+        if (blockHasBottomPadding && opt === 'never') {
+            errors.push({
+                start: lastToken,
+                end: tokenAfterLast,
+                message: message,
+                fix: removeEmptyLinesBetweenTokens(lastToken, tokenAfterLast)
+            });
+        }
+
+        if (!blockHasBottomPadding && opt === 'always') {
+            errors.push({
+                start: lastToken,
+                end: tokenAfterLast,
+                message: message,
+                fix: function (fixer) {
+                    return fixer.insertTextAfter(lastToken, '\n');
+                }
+            });
+        }
+
+        return errors;
+    };
+}
+
+module.exports = {
+    meta: {
+        docs: {
+            description: 'Enforce spacing between test functions',
+            category: 'Stylistic Issues',
+            recommended: false
+        },
+        fixable: 'whitespace',
+        schema: [ {
+            oneOf: [ {
+                enum: [ 'always', 'never' ]
+            } ]
+        } ]
+    },
+    create: function (context) {
+        var option = context.options[0] || 'always',
+            message = option === 'never'
+                ? 'Test functions must not be padded by blank lines.'
+                : 'Test functions must be padded by blank lines.',
+            sourceCode = context.getSourceCode(),
+            topChecker = getTopChecker(sourceCode, option, message),
+            bottomChecker = getBottomChecker(sourceCode, option, message),
+            errors = [];
+
+        return {
+            'Program:exit': function () {
+                R.pipe(
+                  R.uniqWith(function (a, b) {
+                      return a.start.start === b.start.start && a.end.end === b.end.end;
+                  }),
+                  R.map(function (error) {
+                      return {
+                          message: error.message,
+                          loc: {
+                              start: error.start.loc.end,
+                              end: error.end.loc.start
+                          },
+                          fix: error.fix
+                      };
+                  }),
+                  R.forEach(function (error) {
+                      context.report(error);
+                  })
+                )(errors);
+            },
+            ExpressionStatement: function (node) {
+                var tokenBefore = sourceCode.getTokenBefore(node),
+                    tokenAfter = sourceCode.getTokenAfter(node);
+
+                if (!tokenBefore || !tokenAfter || !astUtil.isTestFunction(node.expression)) {
+                    return;
+                }
+
+                errors = R.reduce(
```

Comment:
Is the usage of `R.reduce` really necessary here? I think `errors = R.concat(errors, [ topChecker(node), bottomChecker(node) ])` should be sufficient.

Reply:
That's what I tried initially but `R.concat` has arity of 2 - it doesn't care about subsequent parameters.  When passing an array like that it turns it into

```
[
  [
    {top-error},
    {bottom-error},
  ],
  //...etc...
]
```

I stumped across that pattern here: https://stackoverflow.com/a/34376699.

Reply:
Ah, you are right, that’s because `topChecker` and `bottomChecker` returns an array as well.